[CI] ReindexManagementClientYamlTestSuiteIT test {yaml=reindex/30_cancel_reindex/Cancel running reindex returns response and GET confirms completed} failing

**Build Scans:**
- [elasticsearch-intake #34354 / part1](https://gradle-enterprise.elastic.co/s/scatsm23vyh3q)
- [elasticsearch-pull-request #118392 / part-1](https://gradle-enterprise.elastic.co/s/2ycrynnaqckpu)
- [elasticsearch-pull-request #118171 / part-1](https://gradle-enterprise.elastic.co/s/mcmbdpt3o7tei)
- [elasticsearch-pull-request #118057 / part-1](https://gradle-enterprise.elastic.co/s/d4l4dpsklwgde)

**Reproduction Line:**
```
./gradlew ":modules:reindex-management:yamlRestTest" --tests "org.elasticsearch.reindex.management.ReindexManagementClientYamlTestSuiteIT.test {yaml=reindex/30_cancel_reindex/Cancel running reindex returns response and GET confirms completed}" -Dtests.seed=B99B4680CAF89268 -Dtests.locale=en-PG -Dtests.timezone=America/Jamaica -Druntime.java=25
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.reindex.management.ReindexManagementClientYamlTestSuiteIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('test%20%7Byaml%3Dreindex%2F30_cancel_reindex%2FCancel%20running%20reindex%20returns%20response%20and%20GET%20confirms%20completed%7D'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: java.lang.AssertionError: field [status.canceled] is null
```

**Issue Reasons:**
- [main] 4 failures in test test {yaml=reindex/30_cancel_reindex/Cancel running reindex returns response and GET confirms completed} (0.5% fail rate in 753 executions)
- [main] 3 failures in step part-1 (0.7% fail rate in 440 executions)
- [main] 3 failures in pipeline elasticsearch-pull-request (0.7% fail rate in 434 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/modules/reindex-management/src/main/java/org/elasticsearch/reindex/management/TransportCancelReindexAction.java
+++ b/modules/reindex-management/src/main/java/org/elasticsearch/reindex/management/TransportCancelReindexAction.java
@@ -19,12 +19,14 @@ import org.elasticsearch.action.TaskOperationFailure;
 import org.elasticsearch.action.admin.cluster.node.tasks.cancel.CancelTasksRequest;
 import org.elasticsearch.action.support.ActionFilters;
 import org.elasticsearch.action.support.tasks.TransportTasksProjectAction;
+import org.elasticsearch.client.internal.Client;
 import org.elasticsearch.cluster.project.ProjectResolver;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.index.reindex.BulkByScrollTask;
 import org.elasticsearch.injection.guice.Inject;
 import org.elasticsearch.tasks.CancellableTask;
 import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.tasks.TaskInfo;
 import org.elasticsearch.tasks.TaskResult;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.TransportService;
@@ -40,12 +42,15 @@ public class TransportCancelReindexAction extends TransportTasksProjectAction<
 
     public static final ActionType<CancelReindexResponse> TYPE = new ActionType<>("cluster:admin/reindex/cancel");
 
+    private final Client client;
+
     @Inject
     public TransportCancelReindexAction(
         final ClusterService clusterService,
         final TransportService transportService,
         final ActionFilters actionFilters,
-        final ProjectResolver projectResolver
+        final ProjectResolver projectResolver,
+        final Client client
     ) {
         super(
             TYPE.name(),
@@ -57,6 +62,7 @@ public class TransportCancelReindexAction extends TransportTasksProjectAction<
             transportService.getThreadPool().executor(ThreadPool.Names.GENERIC),
             projectResolver
         );
+        this.client = client;
     }
 
     @Override
@@ -77,16 +83,29 @@ public class TransportCancelReindexAction extends TransportTasksProjectAction<
     ) {
         assert task instanceof BulkByScrollTask : "Task should be a BulkByScrollTask";
 
+        // cancel the task asynchronously, and if waitForCompletion=true, then wait for it to finish to have the full correct response.
         taskManager.cancelTaskAndDescendants(
             task,
             CancelTasksRequest.DEFAULT_REASON,
-            request.waitForCompletion(),
-            ActionListener.wrap(ignored -> {
-                final TaskResult completedTaskResult = request.waitForCompletion()
-                    ? new TaskResult(true, task.taskInfo(clusterService.localNode().getId(), true))
-                    : null;
-                listener.onResponse(new CancelReindexTaskResponse(completedTaskResult));
-            }, listener::onFailure)
+            false,
+            listener.delegateFailureAndWrap((cancelListener, r) -> {
+                if (request.waitForCompletion()) {
+                    final TaskId taskId = new TaskId(clusterService.localNode().getId(), task.getId());
+                    final GetReindexRequest getRequest = new GetReindexRequest(taskId, true, null);
+                    client.execute(
+                        TransportGetReindexAction.TYPE,
+                        getRequest,
+                        cancelListener.delegateFailureAndWrap(
+                            (l, getResp) -> l.onResponse(
+                                // return cancelled=true. GET will return false since it's not *currently* cancelled.
+                                new CancelReindexTaskResponse(taskResultWithCancelledTrue(getResp.getTaskResult()))
+                            )
+                        )
+                    );
+                } else {
+                    cancelListener.onResponse(new CancelReindexTaskResponse((TaskResult) null));
+                }
+            })
         );
     }
 
@@ -118,5 +137,24 @@ public class TransportCancelReindexAction extends TransportTasksProjectAction<
 
     private static ResourceNotFoundException reindexWithTaskIdNotFoundException(final TaskId requestedTaskId) {
         return new ResourceNotFoundException("reindex task [{}] either not found or completed", requestedTaskId);
+    }
+
+    private TaskResult taskResultWithCancelledTrue(final TaskResult r) {
+        final TaskInfo taskInfo = r.getTask();
+        final TaskInfo newTaskInfo = new TaskInfo(
+            taskInfo.taskId(),
+            taskInfo.type(),
+            taskInfo.node(),
+            taskInfo.action(),
+            taskInfo.description(),
+            taskInfo.status(),
+            taskInfo.startTime(),
+            taskInfo.runningTimeNanos(),
+            taskInfo.cancellable(),
+            true,
+            taskInfo.parentTaskId(),
+            taskInfo.headers()
+        );
+        return new TaskResult(r.isCompleted(), newTaskInfo, r.getError(), r.getResponse());
     }
 }

--- a/modules/reindex-management/src/yamlRestTest/resources/rest-api-spec/test/reindex/30_cancel_reindex.yml
+++ b/modules/reindex-management/src/yamlRestTest/resources/rest-api-spec/test/reindex/30_cancel_reindex.yml
@@ -53,8 +53,7 @@ setup:
   - match: { cancelled: true }
   - match: { status.canceled: "by user request" }
   - gte: { status.total: 0 }
-  # different from `GET _reindex`
-  - not_exists: response
+  - exists: error
 
   - do:
       reindex_get:

--- a/modules/reindex-management/src/yamlRestTest/resources/rest-api-spec/test/reindex/30_cancel_reindex.yml
+++ b/modules/reindex-management/src/yamlRestTest/resources/rest-api-spec/test/reindex/30_cancel_reindex.yml
@@ -53,7 +53,7 @@ setup:
   - match: { cancelled: true }
   - match: { status.canceled: "by user request" }
   - gte: { status.total: 0 }
-  - exists: error
+  # depending on how quickly we cancel, we might either get an error (shard failure) or response, so not asserting
 
   - do:
       reindex_get:

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -335,9 +335,6 @@ tests:
 - class: org.elasticsearch.xpack.gpu.GPUPluginInitializationWithGPUIT
   method: testAutoModeSupportedVectorType
   issue: https://github.com/elastic/elasticsearch/issues/142072
-- class: org.elasticsearch.reindex.management.ReindexManagementClientYamlTestSuiteIT
-  method: test {yaml=reindex/30_cancel_reindex/Cancel running reindex returns response and GET confirms completed}
-  issue: https://github.com/elastic/elasticsearch/issues/142079
 - class: org.elasticsearch.search.aggregations.metrics.TopHitsIT
   method: testMixedSortFieldTypes
   issue: https://github.com/elastic/elasticsearch/issues/142077

